rust-install/src/mock/clitools.rs

```diff
@@ -0,0 +1,392 @@
+//! A mock distribution server used by tests/cli-v1.rs and
+//! tests/cli-v2.rs
+
+use std::path::{PathBuf, Path};
+use std::env;
+use std::process::Command;
+use std::env::consts::EXE_SUFFIX;
+use std::fs::{self, File};
+use std::io::Read;
+use std::sync::Mutex;
+use tempdir::TempDir;
+use mock::{MockInstallerBuilder, MockCommand};
+use mock::dist::{MockDistServer, MockChannel, MockPackage,
+                 MockTargettedPackage, MockComponent, change_channel_date,
+                 ManifestVersion};
+use dist::ToolchainDesc;
+use utils;
+use hyper::Url;
+use scopeguard;
+
+/// The configuration used by the tests in this module
+pub struct Config {
+    /// Where we put the multirust / rustc / cargo bins
+    pub exedir: TempDir,
+    /// The distribution server
+    pub distdir: TempDir,
+    /// MULTIRUST_HOME
+    pub homedir: TempDir,
+}
+
+/// Run this to create the test environment containing multirust, and
+/// a mock dist server.
+pub fn setup(vs: &[ManifestVersion], f: &Fn(&Config)) {
+    // Unset env variables that will break our testing
+    env::remove_var("MULTIRUST_TOOLCHAIN");
+
+    let ref config = Config {
+        exedir: TempDir::new("multirust").unwrap(),
+        distdir: TempDir::new("multirust").unwrap(),
+        homedir: TempDir::new("multirust").unwrap(),
+    };
+
+    create_mock_dist_server(&config.distdir.path(), vs);
+
+    let current_exe_path = env::current_exe().map(PathBuf::from).unwrap();
+    let exe_dir = current_exe_path.parent().unwrap();
+    let ref multirust_build_path = exe_dir.join(format!("multirust-rs{}", EXE_SUFFIX));
+
+    let multirust_path = config.exedir.path().join(format!("multirust{}", EXE_SUFFIX));
+    let rustc_path = config.exedir.path().join(format!("rustc{}", EXE_SUFFIX));
+    let rustdoc_path = config.exedir.path().join(format!("rustdoc{}", EXE_SUFFIX));
+    let cargo_path = config.exedir.path().join(format!("cargo{}", EXE_SUFFIX));
+
+    fs::copy(multirust_build_path, multirust_path).unwrap();
+    fs::copy(multirust_build_path, rustc_path).unwrap();
+    fs::copy(multirust_build_path, rustdoc_path).unwrap();
+    fs::copy(multirust_build_path, cargo_path).unwrap();
+
+    f(config);
+}
+
+/// Change the current distribution manifest to a particular date
+pub fn set_current_dist_date(config: &Config, date: &str) {
+    let ref url = Url::from_file_path(config.distdir.path()).unwrap();
+    for channel in &["nightly", "beta", "stable"] {
+        change_channel_date(url, channel, date);
+    }
+}
+
+pub fn expect_ok(config: &Config, args: &[&str]) {
+    expect_stdout_ok(config, args, "");
+}
+
+pub fn expect_err(config: &Config, args: &[&str], expected: &str) {
```

Comment:
We may want to consider unifying the test suites here eventually with Cargo as lots of this is already implemented there as well, but not that critical for now